training/creating_dataset.py

```diff
@@ -42,13 +83,149 @@ def is_path_correct(path: str) -> bool:
+        and "cantrip" not in i.get("system").get("traits").get("value")
+        and i.get("system").get("level").get("value") == spell_level
+    ]
+
+    return len(spells)
+
+
+def count_damage_expected_value(damage_dict: dict[dict]) -> float:
+    """
+    Calculate the total expected value of damage based on a dictionary of damage specifications.
+    :param damage_dict: A dictionary where keys represent different sources of damage,
+                        and values are dictionaries with the "damage" key containing damage specifications.
+                        Damage can be a constant value or a dice roll in the format 'NdM', 'NdM+X', or 'NdM-X',
+                        where N is the number of dice, M is the number of sides on the dice, and X is an optional
+                        positive or negative constant.
+    :return: The calculated total expected value of damage.
+    """
+    total_expected_val = 0
+
+    # chance that one melee item have multiple damage types
+    for key, value in damage_dict.items():
+        damage = value.get("damage")
+
+        if "d" not in damage:
+            # constant damage value
+            total_expected_val += int(damage)
+            continue
+        # split dice roll
+        roll_nr, dice_type = damage.split("d")
+        add = 0
+        if "+" in dice_type:
+            # get possible positive additional value for damage
+            dice_type, add = dice_type.split("+")
+            add = int(add)
+        if "-" in dice_type:
+            # get possible negative additional value for damage
+            dice_type, add = dice_type.split("-")
+            add = -int(add)
+
+        roll_nr, dice_type = int(roll_nr), int(dice_type)
+        # count expected value with additional damages
+        total_expected_val += roll_nr * (dice_type + 1) / 2 + add
+
+    return total_expected_val
+
+
+def get_max_melee_bonus_damage(
+    items_list: list[dict], weaponType: str
+) -> tuple[int, float]:
+    """
+    Function used for pd.Series.apply()\n
+    Get the maximum damageRoll bonus and associated damage from a list of melee of a specific weaponType.
+    :param items_list: A list of dictionaries representing melee weapons, each with relevant attributes.
+    :param weaponType: The type of weapon to filter by.
+    :return: A tuple containing the maximum bonus and the calculated damage associated with that bonus.
+             If no matching melee weapons are found, returns (0, 0).
+    """
+    melee = [
+        i.get("system")
+        for i in items_list
+        if i.get("type") == "melee"
+        and i.get("system").get("weaponType").get("value") == weaponType
+    ]
+
+    if not melee:
+        return 0, 0
+
+    max_bonus_indx, max_bonus = max(
+        [(inx, val.get("bonus").get("value")) for inx, val in enumerate(melee)],
+        key=lambda x: x[1],
+    )
+
+    return max_bonus, count_damage_expected_value(
+        melee[max_bonus_indx].get("damageRolls")
+    )
```

Comment:
1. Lepiej trzymać się jednego skrótu na "index". Albo `idx` (polecam, najczęstsze z mojego doświadczenia), albo `indx`. `inx` to raczej literówka
2. Uporządkowałbym to dla czytelności. Najpierw tworzymy pary `(idx, val)`, potem max, potem wybieramy element dla indeksu, potem liczymy wartość oczekiwaną, każde w osobnej linijce. Bo obecnie jest tu strasznie gęsto z tym, co się dzieje, w 2 linijki kodu (które są de facto tak długie, że się rozbijają na kilka)